File: 2_Add_Two_Numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode head, curr, nxt, nxtnxt;
        int sum, currVal, carry, l1Val, l2Val;
        boolean hasCarry;
        carry = 0;
        
        // Initialize the first node as head
        sum = l1.val + l2.val;
        currVal = sum % 10;
        hasCarry = sum >= 10;
        if(hasCarry) carry = sum / 10;
        head = curr = new ListNode(currVal);
        
        l1 = l1.next;
        l2 = l2.next;
        
        // Iteratively sum the remaining nodes
        while(l1 != null || l2 != null){
            l1Val = (l1 != null) ? l1.val : 0;
            l2Val = (l2 != null) ? l2.val : 0;
            
            // Calculate the value for the new node
            sum = l1Val + l2Val;
            if(hasCarry) sum += carry;
            currVal = sum % 10;
            hasCarry = sum >= 10;
            if(hasCarry) carry = sum / 10;
            
            // Create and link the next node
            nxt = new ListNode(currVal);
            curr.next = nxt;
            curr = nxt;
            l1 = (l1 != null) ? l1.next : null;
            l2 = (l2 != null) ? l2.next : null;
        }
        // Link last node if there is a carry
        if(hasCarry) curr.next = new ListNode(carry);
        
        return head;
    }
}

File: 2_Add_Two_Numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        # Assign first node sum
        l1Ptr = l1
        l2Ptr = l2
        currSum = l1Ptr.val + l2Ptr.val
        currVal = currSum % 10
        hasCarry = currSum >= 10
        if hasCarry: carry = currSum // 10
        
        head = currNode = ListNode(currVal)
        l1Ptr = l1Ptr.next
        l2Ptr = l2Ptr.next
        
        while(l1Ptr is not None or l2Ptr is not None):
            # Assign value and advance pointer if end of list not reached
            # otherwise a null node from one list contributes 0 to the curr sum
            if(l1Ptr is not None): 
                leftVal = l1Ptr.val
                l1Ptr = l1Ptr.next
            else: 
                leftVal = 0 
            if(l2Ptr is not None): 
                rightVal = l2Ptr.val
                l2Ptr = l2Ptr.next
            else: 
                rightVal = 0
            
            # Calculate value of current node, adding previous carry if applicable
            currSum = leftVal + rightVal
            if(hasCarry): currSum += carry              
            currVal = currSum % 10
            
            # Assign carry for next node if relevant
            hasCarry = currSum >= 10
            if(hasCarry): carry = currSum // 10
            
            nextNode = ListNode(currVal)
            currNode.next = nextNode
            currNode = nextNode
        
        # Last node(s) contributed a carry sum past the end of the list, add it as the last node
        if(hasCarry):
            currNode.next = ListNode(carry)
        
        return head
            

        
        